elastic/kibana
# [Enterprise Search][tech debt] Refactor external URL generation to a single DRY helper

## Summary

This PR stems from [feedback comment from Scotty](https://github.com/elastic/kibana/pull/74705#discussion_r468196769) around DRYing out App Search's static route strings into a `routes` file, as well as creating helpers to generate external/legacy URLs in the front-end.

I decided to create a single shared URL helper that generates AS & WS links. These URLs are intended for the front-end and updates accordingly if we detect that the Enterprise Search instance (e.g. Elastic Cloud) is using public/vanity `external_url`s.

I also took the opportunity to clean up `public/plugin.ts` a bit - the publicUrl method/var names are now a little more generic (in anticipation of more initial data work we'll be doing shortly), and we're no longer mutating our `config` obj (hurray).

TL;DR changelist:

- Destructure `externalUrl.getAppSearchUrl()/.getWorkplaceSearchUrl()` from `KibanaContext` from now on to generate external links (replaces Workplace Search's `useRoutes.getWSRoute`)
- The old `enterpriseSearchUrl` context value is gone - you can use `externalUrl.enterpriseSearchUrl` if you need to generate your own super custom URL for some reason. You can also use `config.host` if you need the non-vanity host URL also for some reason (although you shouldn't really).
- New `routes.ts` file for App Search
- Some plugin.ts setup for our upcoming initialData rework

## QA/testing

This is one of those tech debt PRs where nothing should functionally have changed, and for this to be working correctly, everything should work the same as before.

- App Search
  - [x] Nav links still work as expected
  - [x] Engine table links still work as expected 
  - [x] "Launch App Search" header button still works as expected
  - [x] Empty state "create first engine" button still works as expected
- Workplace Search
  - [x] All Overview links still work as expected
  - [x] "Launch Workplace Search" header button still works as expected
  - [x] Nav links still work as expected (go to /groups to test this) 
- Both
  - [x] When `enterpriseSearch.host` is not set in `kibana.yml`, the setup guide redirect/routing works as expected
  - [x] Error state: `config.host` URL shows correct URL

### Checklist
